File: 0141-linked-list-cycle/0141-linked-list-cycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func hasCycle(head *ListNode) bool {
    nodeSet := map[*ListNode]int{}
    cur := head
    
    for cur != nil {
        if _, ok := nodeSet[cur]; ok {
            return true
        }
        nodeSet[cur]++
        
        cur = cur.Next
    }
    return false
}
/*
func hasCycle(head *ListNode) bool {
    if head == nil {
        return false
    }
    
    slow := head
    fast := slow.Next
    
    for fast != nil {
        if fast == slow { 
            return true
        }
        
        fast = fast.Next 
        if fast == nil {
            return false
        }
        
        if fast == slow { 
            return true
        }
        
        slow = slow.Next 
        fast = fast.Next
    }
    
    return false
}
*/

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

"""
class Solution(object):
    def hasCycle(self, head):
        visited = {}
        while head != None:
            if visited.get(head, False):
                return True
            visited[head] = True
            head = head.next
        return False    
"""

class Solution(object):
    def hasCycle(self, head):
        fast, slow = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if fast == slow:
                return True
        return False    